### DISCUSSION TITLE
Feat: Room + User metadata + Audio level indication (X is speaking)

### DISCUSSION THREAD

**Discussion by jibon57** [8xFF/atm0s-media-server#358] [Ideas]
First of thanks for the excellent project!!

I was going through the code & did not find option for user metadata. Although I saw room metadata but from JS SDK default option set to undefined. I think room & user metadata both are important option to have. Also option to update those info & send event on update to every connected user. You can consider this during your development.  

Audio level indication will be another feature to have by this way UI can be updated based on the current talking user.

Thank you!

**Comment by giangndm** [comment]
@jibon57 yes, metadata is very important for integrating with other system or in UI/UX. I will check if it easy to implement I will do it next week.

**Reply by jibon57** [reply]
Thank you for your quick response. I'm also bit lost to saw different sdk on JS for example: 

https://github.com/8xFF/atm0s-media-sdk-ts
https://github.com/8xFF/atm0s-media-sdk-js

Which one to follow? Also is there any server related SDK/API? For example: Create room (with name, metadata, permission etc), list rooms with info/status, delete/end room etc.. ?

**Reply by giangndm** [reply]
@jibon57 https://github.com/8xFF/atm0s-media-sdk-ts is currently updating, [atm0s-media-sdk-js](https://github.com/8xFF/atm0s-media-sdk-js) is deprecated for old-legacy media-server

**Reply by giangndm** [reply]
@jibon57 I just checked. Media-server already implemented but ts-sdk missing config. I just create a small patch and merged into ts-sdk. Can you check if it can work in your case?

https://github.com/8xFF/atm0s-media-sdk-ts/pull/10

**Reply by jibon57** [reply]
Thanks @giangndm that's excellent. But do you think updating/adding that from client side will be safe? Because in metadata may contain certain secret info. I still prefer to manage those from sever level API. Having client side is Ok but need to ensure that user can't manage it my themselves as token will be always visible.

**Reply by jibon57** [reply]
At least room metadata should not be updated/added from client side.

**Reply by giangndm** [reply]
@jibon57 Oh, I see. So we need a other level of metadata, which should generated belong with join token. It only need some small change so I am going to implement it next week.

**Reply by jibon57** [reply]
Perfect! Thanks

**Reply by jibon57** [reply]
Another issue, at present by joining the session room are creating which have one security issue. For example ttl of token is 1 hour & after 30 minutes admin had ended meeting but as the token can be reused and any user can start new session. This will be hard to manage. Meeting/session should be created my server level API only. Later using token user will be join only with limited authorization. You can think about it. Or at least one option to disable room creation just by joining token. Livekit also have this option which is I think good.

**Comment by giangndm** [comment]
The question is how we do with metadata, is is necessary to dynamic for allowing client update on-air

**Reply by jibon57** [reply]
I believe having separate server related API will be best in this case. Same as [livekit](https://github.com/livekit/livekit) is doing now

**Reply by giangndm** [reply]
@jibon57 can you send me the document of livekit about the feature?

**Reply by jibon57** [reply]
https://docs.livekit.io/realtime/server/generating-tokens/
https://docs.livekit.io/realtime/server/managing-rooms/
https://docs.livekit.io/realtime/server/managing-participants/

**Comment by giangndm** [comment]
@jibon57 about audio activity, you can check here:

https://github.com/8xFF/atm0s-media-sdk-ts/blob/main/packages/sdk-react-hooks/src/hooks/consumer.tsx
https://github.com/8xFF/atm0s-media-sdk-ts/blob/main/packages/sdk-react-hooks/src/hooks/mixer.tsx

**Reply by jibon57** [reply]
thanks

**Comment by giangndm** [comment]
@jibon57 I just merged userdata feature, which is embedded into peer token.

https://github.com/8xFF/atm0s-media-sdk-ts/pull/11
https://github.com/8xFF/atm0s-media-server/pull/379

Can you give a try?

**Reply by jibon57** [reply]
wow!! nice! Do you think the name is bit confusing? Only metadata or roomMetadata or roomExtraData etc could be better? As that's not userData.

**Reply by giangndm** [reply]
Ohh, I see. because it is peer level then is it peerExtraData make sense? metadata is already used for data set by client.

**Reply by jibon57** [reply]
Yes, that will nice too.

**Reply by giangndm** [reply]
I just created PR for rename it to extra_data. It will be merged soon.

https://github.com/8xFF/atm0s-media-sdk-ts/pull/15
https://github.com/8xFF/atm0s-media-server/pull/386
